### ISSUE TITLE
Text Chat Support

### ISSUE THREAD

**Issue by VLKAlexander** [Neos-Metaverse/NeosPublic#1914]
Hello! I'm glad to see that development on desktop support has progressed, but I have noticed a lack of text chat support for the platform. While I understand that there are third-party tools for text chat, these are not easily accessible to new users, and many new players have to immediately ask how to find them, or if they even exist. 

An official, easily accessible text chat would be user-friendly, and allow for users without microphones, without desire to speak using voice chat, users physically cannot speak, and users that cannot hear to communicate with others easily. There are a large variety of use-cases for an official text chat function.

Function-wise, text chat would be limited to the world one inhabits. For desktop users, pressing the Enter Key or another alternative would allow them to access the text chat window. The user would type their message and press the Enter key again to send and close it. Not all VR users may need a text chat, but many VR platforms have a builtin system keyboard system already.

VR users can choose to either see speech bubbles appear over text chat user's avatars, have the world text chat located in one corner of their screen, or hide the text chat entirely. They can open text chat using a controller shortcut, such as clicking down on one of the sticks. 

This system could function near-identically to the existing mute helper, but accessible by default via keyboard shortcut for all users. This saves time in the future, as new users will not have to ask for the location of the text chat each time a player joins the game, and it would be accessible across each of their avatars automatically.

**Comment by TehTurk** [comment]
So the issue with this is that it's not strictly a desktop environment where this would benefit. The reason text chat doesn't work well in VR environments is due to you have alot of others things to consider, body language, social cues, etc. Text chat tools have been made but they will always be clunky. You need to consider all players when making something as such.

**Comment by VLKAlexander** [comment]
Adding a universal text chat would take into consideration that not all players have access to microphones, can speak, can hear, or may not wish to use their voice to communicate. VR users may choose to use voice chat if they want to use things like body language to communicate, but desktop users already do not have access to this. The argument that text chat should not be available to users by virtue of not all VR users desiring it is ultimately not considering that many people on desktop do not or can not use voice chat. 

Some features for desktop users are expected to be clunky as well, given that VR allows for more freedom of movement. It is not wrong to have a feature that is more easily usable by desktop users than VR ones.

**Comment by TehTurk** [comment]
Your misunderstanding my response. Your asking for a text chat, but what your looking for is a mute helper addition or add on in some form not a dedicated chat. Having a chat window in VR can be neat or cool, but it's not thinking for the future or considering the differences in communication as well too. It's not engaging or personal to just stare at someone and have them just spit words at in a text format. 

I understand the issue of accessibility and ease of use, but there are some lines of where benefiting desktop over VR not a design goal. Look at any other SocialVR Platform, do they have any form of text chat? Most message formats are tied to messages not a general chat window.

**Comment by VLKAlexander** [comment]
Users should not have to ask for help each time they want to find the text chat function for the platform. If users had to use an addon in their inventory to move around, they may be dissuaded from using the platform or argue in favor of simply having movement controls available to them in-game. 

I have no doubt that Neos will serve as a gateway for individuals on desktop to buy into VR in favor of a more immersive experience, however, the vast majority of desktop users do not have access to VR, and many, due to privacy concerns, disability, and ease of use, may prefer to use text chat instead. 

Desktop already benefits in a variety of ways over VR, and this is part of why desktop support was added. The vast majority of Steam users favor desktop over VR due to price and accessibility, and the vast majority of Steam games and existing virtual worlds have text chat functions. Desktop will also benefit heavily insofar that it will be easier for users to create and edit content, given that the vast majority of 3D artists do not operate in VR. Neos has already set a precedent for some functions being better suited to VR users and others being more suited to VR ones. 

As I understand it, Neos is a competitor to VRChat, therefore, it is inarguable that, because VRChat does not have text chat, Neos should not have either. VRChat lacks a great many things that Neos has, such as more accessible build and creative tools, easier world creation, and a more structured and complex node-based scripting language. Using VRChat as an example of how Neos should be would only detract from the endless possibilities of Neos.

**Comment by TehTurk** [comment]
No they shouldn't and it should be default in some degree, but as I said earlier in the discord chat your asking for something without also considering what the issue is. Your giving us a solution, not the problem. 

Gateways aside I'm not focused on what Desktop does for the platform in this issue, I'm focused on what demanding a feature like this will do to the rest of the platform. Your only considering desktop in this not VR. That's the problem. Especially when the focus has been VR First, and always will be. 

Having a solid desktop mode is key sure, but how it's implemented also matters greatly too. Vocal Social Conversations vastly differ from a text to voice conversation, or even other cues that can occur. There's a difference a phone call vs a text chat have different feelings or implications tied to them. Neos has always been focused on the collaboration so it's going to focus with that in mind.

I'm not going to argue about this because your going into semantics at this point too so I'm gonna boil down my points to simple Yes Nos so you understand we are on the same page, but the way your asking/going is being presented poorly:
-Should it be a thing? Yes
-Should it be a regular Text Chat? No
-Should it consider both VR and Desktop users? Yes
-Should it be an enhanced Mute Helper? Yes

If this issue was called "Official Mute Helper" I'd be all for it maybe reword it so it's a issue that's occurring not a feature request, You also skipped the templates so it shows your not really working within the process to some degree, but what your asking is for a text chat comparable to an MMO from yester-year. You also don't consider how uncomfortable it is to have texture chat always in your Fov in VR, or where it can be problematic to manage it's position or usage. 

If It was a Mute Helper, it's better because I'm actually looking at the person, I'm actually speaking to the person, the focus is on them. Presence matters greatly in VR platforms.

**Comment by AftrLite** [comment]
What about, like.. Military veterans with damaged hearing
wouldn't it be easier for them to use an integrated, easily-accessible text chat feature? .__.

**Comment by shadowpanther** [comment]
Those veterans and other users who have trouble hearing would benefit more from speech-to-text in some form, because most other users around them would continue to speak as they were. Imagine having YouTube-like auto subtitles for social interactions.

**Comment by TehTurk** [comment]
That's a completely valid reason, but that would have to do more with accessibility settings more then say a text chat. I've been having chats with users in the space for accessibility so figuring that out for VR is also tricky too. That's a bigger conversation and problem that'll definitely need a lot of proper thought behind it.

**Comment by Nvlt** [comment]
You know the only valid reason you need is that there is a large number of people who don't want to voice, and there are also a large number of users myself included that are hearing impaired. All you are doing is alienating your user base needlessly, which is why I currently don't even use your app, I'm waiting for it to be good enough to use.

Though don't confuse what I'm saying as asking for just a accessibility feature for only the group that needs it.

Do you need a mute helper? No. I don't mean to be rude but you are over complicating a very simple issue while simultaneously neutering your own "accessibility feature".

Tell me what the hell is even the point of adding it as an accessibility feature if the only ones who enable it are the people who have the disabilities or don't want to voice. So a deaf person can only chat with other deaf people or people who turned on the feature because they wanted it?

If everybody can't see what they type without checking a box in your accessibility menu, then it's not accessible.

What this boils down to is blatant ableism in my opinion.

What you need is an innovative way to type with controllers and a physical keyboard that makes use of hand tracking once that becomes more available, after a basic chat, maybe innovate it over time and add like a vr equivalent to swipe or something, and it needs to be in the default menu and part of the new user introduction.

Not an accessibility check box.

Btw I'm a vr user, not a pc user.

And I just want to say that alienating new users is not helpful to your success at all. 95% of unhappy customers silently leave and never come back and tell 9 to 15 people that your product is shit.

**Comment by H3BO3** [comment]
@Nvlt Desktop mode was only announced as playable a few days ago. We acknowledge text chat is a wanted feature but there's a lot of work that has to be put in for a system like that, especially if it's to work for everyone across VR and desktop. "Make a basic chat" is not feasible, work would be put into something shoddy that would need to be reworked in the future, wasting precious development time. Nobody knows how such a system would work or be laid out yet.

Accusing developers of ableism because of differing priorities is not productive in the least. There are many more features that could be added for accessibility, but some things are already implemented, such as user-requestable badges to indicate deafness or color blindness.

By the way, Turk is not a developer, just a collaborator on the github. Nobody knows what a native text chat system would look like yet, he is explaining the current capabilities and mitigations, and speculating on what the best solution might be.

**Comment by Electronus** [comment]
@VLKAlexander Exactly! so if its going to be done it should be done right and that takes time. the team wants Text chat but doing it right is priority, for now we have the mute helper and while not perfect works, and will get us by until the official implementation of text chat. 

Currently desktop is still in development and that is what is being focused on, it is simply a matter of the team only being able to do so many things at once, text chat while needed is not something that is critical right now. 

I'd suggest at this point we wait for some input from one of the development team as to what their plans are for implementing text communication natively.

**Comment by VLKAlexander** [comment]
That's awesome! It's good to hear that the team has this planned. I'm more than willing to wait and see in that case, especially given how well desktop support has released already.

**Comment by StarfishHidari** [comment]
I am hard of hearing and helped with the design of mute helper 2, I was working on a mute helper 3.0, but i did quickly realize that designing text chat in vr properly is.. complicated. As an example I wanted to add a chat history log as a frequent issue was people missing messages, but a problem that arose with that was it would become likely that people ignore others using text chat and don't actively engage with them, instead only reading the history leaving the people using text chat feeling left out of conversations.

It's not that people are against adding accessibility features, just designing a good text chat that accommodates all scenarios is complicated.

**Comment by ukilop** [comment]
> people ignore others using text chat and don't actively engage with them, instead only reading the history leaving the people using text chat feeling left out of conversations.

speaking from experience, same thing happens in tabletop sim for example
